ff_control/scripts/opt_ctrl_py_node

```diff
@@ -0,0 +1,498 @@
+
+    @property
+    def k_pos(self):
+        return self.get_parameter('opt_gain_kpos').value
+
+    @property
+    def k_input(self):
+        return self.get_parameter('opt_gain_kinput').value
+
+    @property
+    def k_velo(self):
+        return self.get_parameter('opt_gain_kvelo').value
+
+    @property
+    def T(self):
+        return self.get_parameter('opt_horizon_secs').value
+
+    @property
+    def freq(self):
+        return self.get_parameter('opt_freq').value
+
+    def state_is_ready(self) -> bool:
+        """
+        Check if state is ready.
+
+        :return: True if state is ready, False otherwise
+        """
+        return self._state_ready
+
+    def _state_callback(self, msg: FreeFlyerStateStamped) -> None:
+        """
+        Get called when the first current state measurement comes in.
+        """
+        self._state_stamped = copy.deepcopy(msg)
+
+        if not self._state_ready:
+            self._state_ready = True
+            self.state_ready_callback()
+
+    def set_default_hyperparameters(self) -> None:
+        # Optimization far hyperparameters/gains
+        self.get_logger().info("*** Reset to default hyperparams ***")
+        self._close_controller = False
+        horizon = rclpy.Parameter('opt_horizon_secs', rclpy.Parameter.Type.DOUBLE, 2.667)
+        freq = rclpy.Parameter('opt_freq', rclpy.Parameter.Type.DOUBLE, 3.0)
+
+        kth = rclpy.Parameter('opt_gain_kth', rclpy.Parameter.Type.DOUBLE, 1.0)
+        kpos = rclpy.Parameter('opt_gain_kpos', rclpy.Parameter.Type.DOUBLE, 2.5)
+        kin = rclpy.Parameter('opt_gain_kinput', rclpy.Parameter.Type.DOUBLE, 20.0)
+        kvelo = rclpy.Parameter('opt_gain_kvelo', rclpy.Parameter.Type.DOUBLE, 40.0)
+
+        self.set_parameters([horizon, freq, kth, kpos, kin, kvelo])
+        self.w0, self.cont_nlp_solver, self.lbw, self.ubw, self.lbg, self.ubg = self.init_solver()
+
+        self.timer.cancel()
+        self.timer = self.create_timer(0.333, self.control_loop)
+
+
+    def set_close_hyperparameters(self) -> None:
+        # Optimization close hyperparameters/gains
+        self.get_logger().info("*** Set to close hyperparams ***")
+        self._close_controller = True
+        horizon = rclpy.Parameter('opt_horizon_secs', rclpy.Parameter.Type.DOUBLE, 0.3)
+        freq = rclpy.Parameter('opt_freq', rclpy.Parameter.Type.DOUBLE, 10.0)
+
+        kth = rclpy.Parameter('opt_gain_kth', rclpy.Parameter.Type.DOUBLE, 5.0)
+        kpos = rclpy.Parameter('opt_gain_kpos', rclpy.Parameter.Type.DOUBLE, 10.0)
+        kin = rclpy.Parameter('opt_gain_kinput', rclpy.Parameter.Type.DOUBLE, 3.5)
+        kvelo = rclpy.Parameter('opt_gain_kvelo', rclpy.Parameter.Type.DOUBLE, 20.5)
+
+        self.set_parameters([horizon, freq, kth, kpos, kin, kvelo])
+        self.w0, self.cont_nlp_solver, self.lbw, self.ubw, self.lbg, self.ubg = self.init_solver()
+
+        self.timer.cancel()
+        self.timer = self.create_timer(0.1, self.control_loop)
```

Comment:
Maybe it's worth considering having 2 sets of parameters. Switching params on the fly can be inefficient.